man/ltrace.man
.\" 3.2
.\"  /*******************************************************************\
.\"   * Copyright (C) 1995 Lars J Aas <larsa@colargol.stud.idb.hist.no> *
.\"   * based on documentation by C J Needham <col@imdb.com> 1995,      *
.\"   * permission is granted by the authors to freely distribute       *
.\"   *                        providing no fee of any kind is charged. *
.\"  \*******************************************************************/
.\"
.TH LTRACE 1 "RELEASEDATE" " " "The Internet Movie Database vVERSIONNUMBER"
.SH NAME
ltrace \- display information about all movies a person has worked on.
.SH SYNOPSYS
.B
ltrace <searchmode> "name" [options]
.PP
The options are similar to those for
.BR list (1).
.SH DESCRIPTION
This program displays all the information about each of the movies that someone
has worked on.
.SH SEARCHMODES
.TP
.B \-acr <name>
Search for actor.
.TP
.B \-acs <name>
Search for actress.
.TP
.B \-cast <name>
Search both cast lists.
.TP
.B \-dir <name>
Search for director.
.TP
.B \-write <name>
Search for writer.
.TP
.B \-comp <name>
Search for composer.
.TP
.B \-cine <name>
Search for cinematographer.
.TP
.B \-edit <name>
Search for editor.
.TP
.B \-prodes <name>
Search for production designer.
.TP
.B \-costdes <name>
Search for costume designer.
.TP
.B \-prdcr <name>
Search for producer.
.TP
.B \-misc <name>
Search miscellaneous database.
.TP
.B \-name <name>
Search all filmography databases.
.SH OPTIONS
.TP
.B \-yr
Adds year of release information.
.TP
.B \-mrr
Adds movie ratings report information.
.TP
.B \-pl
Add plot summary.
.TP
.B \-triv
Add movie trivia and formats results.
.TP
.B \-chr
Add character names.
.TP
.B \-aka
List any alternative titles.
.TP
.B \-full
Combines all the options above in this section.
.SH EVEN MORE OPTIONS
.TP
.B \-m
Movies only, ignore TV-series.
.TP
.B \-i
Case insensitive search.
.SH EXAMPLE
This is a good one:
.PP
ltrace -d "Hitchcock, Alfred" -full


docs/ChangeLog
-------Version 3.4 Released---------------------

Sun Mar 29 22:40:37 BST 1998

 * biographies list offset fields increased to 4 bytes following overflow
 * various constants updated

-------Version 3.3c Released---------------------

Sat Apr 26 06:51:19 BST 1997

 * added support for crew completion list
 * added support for distributors list
 * adapted mkdbs for more robust processing of filmography lists
 * added support for new business list tag

-------Version 3.3b Released---------------------

Sat Mar 15 08:09:50 GMT 1997

 * added support for ISO-Latin-1

-------Version 3.3 Released---------------------

Wed Oct 16 22:28:54 BST 1996

 * added support for AKA attributes
 * added support for writers ordering
 * moved misc section to after case in title output
 * added support for business list
 * added support for laserdisc list
 * fixed tagline bug
 * diplays URL of web page for name/title

-------Version 3.2h Released---------------------

Tue Oct  8 14:24:21 BST 1996

 * added support for special effects companies list
 * updated various system constants
 * added new movie link types 'edited into' & 'edited from'
 * added extra checks to avoid title/attribute key file corruption in mkdbs.c

-------Version 3.2f Released---------------------

Mon Jun 24 22:04:46 GMT 1996

 * added support for language list

-------Version 3.2e Released---------------------

Sat Mar 16 08:09:39 GMT 1996

 * added support for alternate versions list
 * added bio keys IT & AT

-------Version 3.2d Released---------------------

Tue Aug 15 23:23:47 GMT 1995

 * switched TitleID to unsigned long to handle >65K titles
 * fixed bug in display of literature and genre data
 * added bio key WN

-------Version 3.2c Released---------------------

Sun May 14 23:00:14 GMT 1995

 * raised MAXFILMOGRAPHIES constant to handle larger no. of entries

-------Version 3.2 Released---------------------

Sun Jan 29 22:40:08 GMT 1995

 * added full support for title searching without year/TV/V info

Tue Jan  3 23:00:45 GMT 1995

 * replaced 'biodefs' file
 * fixed bug when looking for common titles with people using -name

Mon Jan  2 21:52:08 GMT 1995

 * log file support
 * better separation of search/display code

Wed Dec 14 21:43:57 GMT 1994

 * fixed bug in 'title' for miscellaneous section
 * added support for the movie links list

Tue Dec 13 22:06:44 GMT 1994
 
 * noted when cast in credits output
 * fixed name ordering problem for names with (I)/(II) marker
 * display Actor/Actress in list -name where appropriate

Wed Dec  7 21:19:19 GMT 1994

 * added support for tag lines & complete cast list
 * put spaces around "." in cast results
 * standardized number of dashes printed out in 'title' / 'list'

Mon Dec  5 21:19:42 GMT 1994

 * multi-country productions displayed on same line

Wed Nov 30 18:14:16 GMT 1994

 * switched to new IMDB license
 * updated 'lfetch' to get new lists
 * fixed "No plot summaries message"
 * simplified MRR processing in mkdb
 * added support for locations & release date lists
 * switched 'Reviews' marker for clarity
 * added support for technical & literature lists

Mon Oct 31 21:42:03 GMT 1994
 
-------Version 3.1 Released---------------------

Mon Oct 17 19:04:26 GMT 1994
 
 * 3.1 code frozen
 * minor bug fixes with lindex

Sun Oct  9 18:37:14 GMT 1994

 * added support for countries list

Tue Oct  4 21:27:53 GMT 1994

 * added support for cast ordering info

Sun Oct  2 23:02:59 GMT 1994

 * added support for color info and sound mix lists
 * added constraint options to 'list'
 * standardized list options

Wed Sep 28 22:22:56 GMT 1994

 * re-ordered title info printing plus separated genres list
 * print genres on one line

Mon Sep 26 21:53:47 GMT 1994

 * added support for production companies list

Sun Sep 25 22:02:07 GMT 1994

 * added constraint options to 'lindex'

Tue Sep  6 20:09:52 GMT 1994

 * added support for certificates & genres list
 * fixed mkballot bug with -ar option

Tue Aug 23 20:08:04 GMT 1994

 * added support for miscellaneous filmographies list

Tue May 24 19:13:47 GMT 1994

 * moviedb-3.0 completed

Thu May 19 22:35:12 GMT 1994

 * switched to merged character names scheme

Tue May 17 21:09:45 GMT 1994

 * new 'mkdb' eliminating need for large array to hold name keys
 * removed duplicate aka titles in 'title -s -aka' queries

Wed May 11 22:11:59 GMT 1994

 * made upperCase a macro for increased speed
 * switched to Phil's new Makefiles

Wed May  4 17:46:08 GMT 1994

 * mapped case of titles/names to match key files
 * added support for running times list
 * lguide error messages

Tue May  3 21:38:30 GMT 1994
 
 * added support for Phil's imoviedb package
 * expanded compression support

Thu Apr 28 19:51:52 GMT 1994

 * finished rewrite to use new database representation

-------Version 2.9 Released---------------------

Wed Mar 16 21:39:40 GMT 1994

 * added support for producers list
 * added support for soundtracks list
 * added support for generic aka lists

Tue Mar 15 23:47:46 GMT 1994

 * replaced -crazy and -triv options with -trivia (also includes goofs & quotes)
 * added support for quotes list
 * added support for goofs list
 * handled -full error messages correctly
 * correct conditions for printing of missing plot summaries message
 * fixed lindex to search all lists when no list options given
 * -mrr no longer implies -m

Mon Mar 14 21:48:21 GMT 1994

 * new version of check-for-program
 * passed $MAKE variable to make subprocesses
 * lfetch modified to match full filenames
 * added 'keepchars' option for CLEANLISTS

-------Version 2.8 Released---------------------

Wed Dec  1 21:58:21 GMT 1993

 * man pages added to package
 * LaTeX documentation added to package

Sat Nov 27 20:53:40 GMT 1993

 * 2.8 code finished!

Thu Nov 25 18:17:00 GMT 1993

 * ported lseen to C

Tue Nov 23 19:16:48 GMT 1993

 * added gendb -votes option

Mon Nov 22 23:54:58 GMT 1993

 * ported lformat to C
 * removed yrf.c file

Wed Nov 17 22:36:44 GMT 1993


src/ltrace.c
/*============================================================================
 *
 *  Program: ltrace.c 
 *
 *  Version: 3.7
 *
 *  Purpose: search list databases for full details of a persons movies
 *
 *  Author:  C J Needham <col@imdb.com>     
 *
 *  Copyright (c) 1996-1999 The Internet Movie Database Inc.
 *
 *  Permission is granted by the copyright holder to distribute this program
 *  is source form only, providing this notice remains intact, and no fee
 *  of any kind is charged. This software is provided as is and there are no 
 *  warranties, expressed or implied.
 * 
 *  Options:
 *    -ar <name>    search actors database
 *    -as <name>    search actress database 
 *    -a <name>     search both cast databases
 *    -d <name>     search directors database
 *    -w <name>     search writers database
 *    -c <name>     search composers database
 *    -ph <name>    search cinematographers database
 *    -ed <name>    search editors database
 *    -pd <name>    search production designers database
 *    -cd <name>    search costume designers database
 *    -pr <name>    search producers database
 *    -misc <name>  search miscellaneous database
 *    -name <name>  search all databases
 *
 *    -mrr        add movie ratings report info to results
 *    -yr         add year of release info & sort chronologically
 *    -chr        add character name info (if available)
 *    -aka        add aka info
 *    -trivia     add trivia, crazy credits, goofs and quotes info
 *    -pl         add plot summaries
 *    -full       combines all options in this group
 *
 *    -m          movies only, ignore TV-series
 *
 *============================================================================
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "log.h"
#include "moviedb.h"
#include "dbutils.h"
#include "filmography.h"
#include "titlesearch.h"
#include "display.h"

#define LTRACE_USAGE1 "usage: ltrace -cast|-acr|-acs|-dir|-write|-comp|-cine|-edit|-prodes|-costdes|"
#define LTRACE_USAGE2 "              -prdcr|-misc|-name <name>] [-mrr -aka -pl -chr -yr -trivia]"
#define LTRACE_USAGE3 "              [-full] [-m]"

int main ( int argc, char **argv )
{
  struct nameSearchRec  *nrec = NULL ;
  struct titleSearchRec *tchain = NULL ;
  struct nameSearchOptRec  noptions = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 } ;
  struct titleSearchOptRec  toptions = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 } ;
  int    trivopts [ NO_OF_TRIV_LISTS ] ;
  int    titleInfoOpts [ NO_OF_TITLE_INFO_LISTS ] ;
  int    i, j, okopt, err = FALSE ;

  logProgram ( argv [ 0 ] ) ;

  if ( argc == 1 )
    err = TRUE ;

  for ( i = 0 ; i < NO_OF_TRIV_LISTS ; i++ )
    trivopts [ i ] = FALSE ;

  for ( i = 0 ; i < NO_OF_TITLE_INFO_LISTS ; i++ )
    titleInfoOpts [ i ] = FALSE ;

  for ( i = 1; i < argc; i++ )
    if ( strcmp ( "-m", argv[i] ) == 0 )
      noptions . mvsonly = TRUE ;
    else if ( strcmp ( "-yr", argv[i] ) == 0 )
    {
      toptions . yropt = YR ;
      noptions . yropt = YR ;
    }
    else if ( strcmp ( "-aka", argv[i] ) == 0 )
    {
      toptions . akaopt = TRUE ;
      noptions . akaopt = TRUE ;
    }
    else if ( strcmp ( "-chr", argv[i] ) == 0 )
    {
      toptions . chopt = TRUE ;
      noptions . chopt = TRUE ;
    }
    else if ( strcmp ( "-mrr", argv[i] ) == 0 )
      toptions . mrropt = MRR ;
    else if ( strcmp ( "-pl", argv[i] ) == 0 )
      toptions . plotopt = TRUE ;
    else if ( strcmp ( "-bus", argv[i] ) == 0 )
      toptions . businessopt = TRUE ;
    else if ( strcmp ( "-ld", argv[i] ) == 0 )
      toptions . ldopt = TRUE ;
    else if ( strcmp ( "-trivia", argv[i] ) == 0 )
    {
      for ( j = 0 ; j < NO_OF_TRIV_LISTS ; j++ )
        trivopts [ j ] = TRUE ;
      for ( j = 0 ; j < NO_OF_TITLE_INFO_LISTS ; j++ )
        titleInfoOpts [ j ] = TRUE ;
    }
    else if ( strcmp ( "-full", argv[i] ) == 0 )
    {
      for ( j = 0 ; j < NO_OF_TRIV_LISTS ; j++ )
        trivopts [ j ] = TRUE ;
      for ( j = 0 ; j < NO_OF_TITLE_INFO_LISTS ; j++ )
        titleInfoOpts [ j ] = TRUE ;
      toptions . yropt = TRUE ;
      toptions . chopt = TRUE ;
      toptions . mrropt = TRUE ;
      toptions . akaopt = TRUE ;
      toptions . plotopt = TRUE ;
      toptions . businessopt = TRUE ;
      toptions . ldopt = TRUE ;
      toptions . litopt = TRUE ;
      toptions . linkopt = TRUE ;
      noptions . yropt = YR ;
      noptions . akaopt = TRUE ;
      noptions . chopt = TRUE ;
      noptions . biopt = TRUE ;
    }
    else
    {
      okopt = FALSE ;
      for ( j = 0 ; j < NO_OF_SEARCH_OPTS ; j++ )
	if ( strcmp ( filmographyOptions [ j ] . option, argv [ i ] ) == 0 || strcmp ( filmographyOptions [ j ] . oldOption, argv [ i ] ) == 0 )
        {
	  okopt = TRUE ;
          if ( ++i < argc )
          {
            nrec = addNameSearchRec ( nrec, argv [ i ], j ) ;
            if ( j == GLOBAL_NAME_SEARCH )
              noptions . searchall = TRUE ;
          }
          else
          {
            err = TRUE ;
	    break ;
          }
	}
      if ( ! okopt )
      {
         (void) fprintf ( stderr, "ltrace: unrecognised option %s\n", argv[i] ) ;
         err = TRUE ;
      }
    }

  if ( err || nrec == NULL )
    moviedbUsage ( LTRACE_USAGE1, LTRACE_USAGE2, LTRACE_USAGE3, NULL, NULL, NULL ) ;

  if ( nrec -> next != NULL )
    moviedbError ( "ltrace: only one name allowed" ) ;

  addNameChainOpts ( nrec, noptions ) ;
  processFilmographySearch ( nrec ) ;
  sortListResults ( nrec ) ;
  tchain = makeTitleChain ( nrec, toptions, trivopts, noptions, titleInfoOpts ) ;
  if ( tchain != NULL )
  {
    processTitleSearch ( tchain ) ;
    displayNameSearchResults ( nrec, TRUE ) ;
    displayTitleSearchResults ( tchain, TRUE ) ;
    freeTitleChain ( tchain ) ; 
  }
  freeNameSearchChain ( nrec ) ; 

  return ( 0 ) ;
}
